### PR TITLE
[SPARK-27128][SQL] Migrate JSON to File Data Source V2

### PR DESCRIPTION
**Pull request by gengliangwang**
## What changes were proposed in this pull request?
Migrate JSON to File Data Source V2

## How was this patch tested?

Unit test

### CODE REVIEW

**Comment by VaibhavDesai** [comment]
**[Test build abc has started](http://google.comtestReport)** for PR 24058 at commit [`aac8841`](https://github.com/apache/spark/commit/aac88410153c287fe56e7ff88ce799f98dec1bbd).

**Comment by VaibhavDesai** [comment]
**[Test build abc has started](http://google.comtestReport)** for PR 24058 at commit [`aac8841`](https://github.com/apache/spark/commit/aac88410153c287fe56e7ff88ce799f98dec1bbd).

**Comment by VaibhavDesai** [comment]
**[Test build abc has finished](http://google.comtestReport)** for PR 24058 at commit [`aac8841`](https://github.com/apache/spark/commit/aac88410153c287fe56e7ff88ce799f98dec1bbd). * This patch merges cleanly.
 * This patch adds no public classes.

**Comment by VaibhavDesai** [comment]
**[Test build abc has started](https://ci-dev.ngcc.nvidia.com/testReport)** for PR 24058 at commit [`aac8841`](https://github.com/apache/spark/commit/aac88410153c287fe56e7ff88ce799f98dec1bbd).

**Comment by VaibhavDesai** [comment]
**[Test build abc has finished](https://ci-dev.ngcc.nvidia.com/testReport)** for PR 24058 at commit [`aac8841`](https://github.com/apache/spark/commit/aac88410153c287fe56e7ff88ce799f98dec1bbd). * This patch merges cleanly.
 * This patch adds no public classes.

**Comment by gengliangwang** [comment]
retest this please.

**Comment by dilipbiswal** [comment]
retest this please

**Comment by gengliangwang** [comment]
retest this please.

**Comment by gengliangwang** [comment]
retest this please.

**Comment by gengliangwang** [comment]
This is ready. Please help review it. @cloud-fan @dongjoon-hyun @HyukjinKwon

File: R/pkg/tests/fulltests/test_sparkSQL.R [hunk 1]
```diff
@@ -3657,8 +3657,7 @@ test_that("Call DataFrameWriter.load() API in Java without path and check argume
   # It makes sure that we can omit path argument in read.df API and then it calls
   # DataFrameWriter.load() without path.
   expect_error(read.df(source = "json"),
-               paste("Error in load : analysis error - Unable to infer schema for JSON.",
-                     "It must be specified manually"))
+               "Error in load : analysis error - Unable to infer schema for")
```
Comment (felixcheung) [inline R/pkg/tests/fulltests/test_sparkSQL.R]: hmm, it doesn't say "JSON" now?

Comment (gengliangwang) [reply R/pkg/tests/fulltests/test_sparkSQL.R]: The current error message is changed to `Unable to infer schema for $tableName`, while the `tableName`is `shortName + path`. I can create another PR to fix that.

**Comment by HyukjinKwon** [comment]
Looks good if this is matched to CSV one. Will take a closer look late in this week

**Comment by gengliangwang** [comment]
retest this please.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/OptimizeMetadataOnlyQuerySuite.scala [hunk 2]
```diff
@@ -139,6 +140,7 @@ class OptimizeMetadataOnlyQuerySuite extends QueryTest with SharedSQLContext {
         val df = spark.read.json(path.getCanonicalPath).select("CoL1", "CoL5", "CoL3").distinct()
         checkAnswer(df, Row("a", "e", "c"))
 
+        df.explain(true)
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/OptimizeMetadataOnlyQuerySuite.scala]: we should remove it

File: sql/core/src/test/scala/org/apache/spark/sql/execution/OptimizeMetadataOnlyQuerySuite.scala [hunk 3]
```diff
@@ -130,7 +130,8 @@ class OptimizeMetadataOnlyQuerySuite extends QueryTest with SharedSQLContext {
   }
 
   test("Incorrect result caused by the rule OptimizeMetadataOnlyQuery") {
-    withSQLConf(OPTIMIZER_METADATA_ONLY.key -> "true") {
+    withSQLConf(OPTIMIZER_METADATA_ONLY.key -> "true",
+      SQLConf.USE_V1_SOURCE_READER_LIST.key -> "json") {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/OptimizeMetadataOnlyQuerySuite.scala]: isn't v2 disabled by default?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala [hunk 4]
```diff
@@ -1961,9 +1961,7 @@ class JsonSuite extends QueryTest with SharedSQLContext with TestJsonData {
         spark.read.schema(schema).json(path).select("_corrupt_record").collect()
       }.getMessage
       assert(msg.contains("only include the internal corrupt record column"))
-      intercept[catalyst.errors.TreeNodeException[_]] {
-        spark.read.schema(schema).json(path).filter($"_corrupt_record".isNotNull).count()
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala]: do we change the behavior for this case?

Comment (gengliangwang) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala]: See the discussion in https://github.com/apache/spark/pull/24005/files#r263881555

File: sql/core/src/test/scala/org/apache/spark/sql/execution/OptimizeMetadataOnlyQuerySuite.scala (hunk 3 above)
Comment (gengliangwang) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/OptimizeMetadataOnlyQuerySuite.scala]: V2 reader is enabled by default.

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**